### PR TITLE
Add documentation for the Kubernetes Kit

### PR DESCRIPTION
**Pull request by MatthewVaadin**
Add documentation for the Kubernetes Kit.

### CODE REVIEW

**Comment by jouni** [comment]
Drop the `-kit` suffix from the folder name, to align with other kits.

File: articles/tools/azure/configuration.asciidoc [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# kubectl get pods
+NAME                                 READY   STATUS    RESTARTS      AGE
+azure-kit-demo-f87bfcbb4-jj5c4       1/1     Running   0             12s
+azure-kit-demo-f87bfcbb4-czkzr       1/1     Running   0             51s
+azure-kit-demo-f87bfcbb4-gjqw6       1/1     Running   0             51s
+azure-kit-demo-f87bfcbb4-rxvjb       1/1     Running   0             51s
+azure-kit-redis-788d56c66-8b259      1/1     Running   0             51s
+----
+
+If you try to increment the counter again, your request is redirected to another pod but the session should be restored and the counter keeps incrementing.
```
Comment (heruan) [inline articles/tools/azure/configuration.asciidoc:40]: This should be removed as it is already covered by the new `getting-started.asciidoc`.

File: articles/tools/azure/update-version.asciidoc [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+4. Make the new version as a default:
+[source,terminal]
+----
+# kubectl apply -f deployment/ingress-v2.yaml
+----
+5. Delete the old version:
+[source,terminal]
+----
+# kubectl delete -f deployment/app-v1.yaml
+----
```
Comment (heruan) [inline articles/tools/azure/update-version.asciidoc:39]: This should be removed as it is already covered by the new `getting-started.asciidoc`.

File: articles/tools/azure/index.asciidoc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+Microsoft Azure is a cloud service provided by Microsoft for application management and related services: software as a service (SaaS); platform as a service (PaaS); and infrastructure as a service (IaaS).
+
+The Vaadin Azure Kit is used to integrate your applications with Azure, to be able to deploy your Vaadin Flow applications to the cloud.
+
+The Azure Kit enables teams to deploy seamlessly and operate Vaadin Java-based applications built using Flow on Azure Cloud. The Kit helps teams quickly to create Vaadin applications that are scalable, highly available, and user-friendly. The Kit is also based on Kubernetes to minimize the risk of cloud-provider lock-in.
+
+Download the latest version from the following link:
+
+http://tools.vaadin.com/nexus/content/repositories/vaadin-prereleases/com/vaadin/azure-kit/[Download Azure Kit,role="button primary water"]
+
```
Comment (heruan) [inline articles/tools/azure/index.asciidoc:22]: I don't think we need a download button for the Kit, since it can be added as a dependency as documented in `getting-started.asciidoc`.

**Review by heruan — CHANGES_REQUESTED** [review]

File: articles/tools/azure/getting-started.asciidoc [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+  <groupId>com.vaadin</groupId>
+  <artifactId>azure-kit-starter</artifactId>
+</dependency>
+----
+
+Add the following to the application configuration file:
+
+.application.properties
+[source,properties]
+vaadin.serialization.include-packages=com.example.application
```
Comment (mcollovati) [inline articles/tools/azure/getting-started.asciidoc:51]: Should we mention to also set the `vaadin.devmode.sessionSerialization.enabled=true` property, to have session serialization debug tool active in development mode?

File: articles/tools/azure/index.asciidoc [hunk 5]
```diff
@@ -10,8 +10,19 @@ section-nav: commercial
 :kit-trial: true
 include::{articles}/_commercial-banner.asciidoc[opts=optional]
 
+Microsoft Azure is a cloud service provided by Microsoft for application management and related services: software as a service (SaaS); platform as a service (PaaS); and infrastructure as a service (IaaS).
```
Comment (gtzluis) [inline articles/tools/azure/index.asciidoc:13]: This looks a bit weird to me, should it instead be:
Microsoft Azure is a cloud service provided by Microsoft for application management and related services: software as a service (SaaS), platform as a service (PaaS), and infrastructure as a service (IaaS).

File: articles/tools/azure/index.asciidoc [hunk 6]
```diff
@@ -10,8 +10,19 @@ section-nav: commercial
...
+
+The Vaadin Azure Kit is used to integrate your applications with Azure, to be able to deploy your Vaadin Flow applications to the cloud.
+
+The Azure Kit enables teams to deploy seamlessly and operate Vaadin Java-based applications built using Flow on Azure Cloud. The Kit helps teams quickly to create Vaadin applications that are scalable, highly available, and user-friendly. The Kit is also based on Kubernetes to minimize the risk of cloud-provider lock-in.
```
Comment (gtzluis) [inline articles/tools/azure/index.asciidoc:17]: it should probably be:
The Kit helps teams to quickly create Vaadin applications that are scalable
and:
cloud-provider -> cloud provider

**Review by gtzluis — COMMENTED** [review]
Some grammar suggestions

File: articles/tools/azure/index.asciidoc [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+= Azure Kit
+
+:commercial-feature: Azure Kit
+:kit-trial: true
+include::{articles}/_commercial-banner.asciidoc[opts=optional]
+
+Microsoft Azure is a cloud service provided by Microsoft for application management and related services: software as a service (SaaS); platform as a service (PaaS); and infrastructure as a service (IaaS).
+
+The Vaadin Azure Kit is used to integrate your applications with Azure, to be able to deploy your Vaadin Flow applications to the cloud.
```
Comment (heruan) [inline articles/tools/azure/index.asciidoc:15]: Let's just use "Azure Kit" without "Vaadin", as we refer also to other kits in the docs and website.
```suggestion
Azure Kit is used to integrate your applications with Azure, to be able to deploy your Vaadin Flow applications to the cloud.
```

File: articles/tools/azure/session-replication.adoc [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+
+A Vaadin application is comprised of UI objects that act as root nodes of graphs of components that represent the web page the user interacts with in through the browser.
+UIs are collected into a [classname]`VaadinSession` that is stored in the HTTP session, to preserve the server-side state across requests.
+For this reason, application code should be carefully written to be serializable for Vaadin session replication to succeed.
+
+.Java serialization documentation
+[TIP]
+For a deeper insight of serialization process, consult https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/index.html[Java documentation] website.
+
+== Vaadin Session Replication Tips
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:55]: Let's ditch "Vaadin" also from here, as it is implicit we're in Vaadin context.
```suggestion
== Session Replication Tips
```

File: articles/tools/azure/session-replication.adoc [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+----
+vaadin.serialization.include-packages=com.example.app.ui,com.example.app.data
+vaadin.serialization.exclude-packages=com.example.app.ui.utils
+----
+
+== Vaadin and HTTP Session
+
+A Vaadin application is comprised of UI objects that act as root nodes of graphs of components that represent the web page the user interacts with in through the browser.
+UIs are collected into a [classname]`VaadinSession` that is stored in the HTTP session, to preserve the server-side state across requests.
+For this reason, application code should be carefully written to be serializable for Vaadin session replication to succeed.
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:49]: We replicate the whole session, not just Vaadin's.
```suggestion
For this reason, application code should be carefully written to be serializable for session replication to succeed.
```

File: articles/tools/azure/session-replication.adoc [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+4. The [classname]`Product` class has a field of type [classname]`Category`.
+Therefore, the [classname]`Category` class must also be made serializable.
+5. The [classname]`Category` class is made serializable by extending [interfacename]`Serializable`.
+
+
+=== Transient Fields for Non-Serializable Objects
+
+If a UI component needs to store a reference to an instance of a not serializable class, it must defined it as `transient`.
+
+In Spring projects, Vaadin Kubernetes Kit is able to detect this fields and re-inject them after deserialization.
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:101]: Remove "Vaadin" also here.
```suggestion
In Spring projects, Kubernetes Kit is able to detect this fields and re-inject them after deserialization.
```

File: articles/tools/azure/session-replication.adoc [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+
+[NOTE]
+In non Spring project, transient fields should be handled manually either by implementing Java serialization hooks (https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/output.html#the-writeobject-method[`writeObject`], https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/output.html#the-writereplace-method[`writeReplace`], https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/input.html#the-readobject-method[`readObject`], https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/input.html#the-readresolve-method[`readResolve`], etc.) or the Vaadin Kubernetes Kit [interfacename]`TransientHandler` interface.
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:104]: And here.
```suggestion
In non Spring project, transient fields should be handled manually either by implementing Java serialization hooks (https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/output.html#the-writeobject-method[`writeObject`], https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/output.html#the-writereplace-method[`writeReplace`], https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/input.html#the-readobject-method[`readObject`], https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/input.html#the-readresolve-method[`readResolve`], etc.) or the Kubernetes Kit [interfacename]`TransientHandler` interface.
```

File: articles/tools/azure/session-replication.adoc [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+        status.setItemLabelGenerator(item -> 
+            service.humanReadableState(item)
+        );
+    }
+}
+----
+
+In this case, probably a solution may be store the `service` reference as [classname]`OrderEditor` transient field, accessing the instance in the lambda with a method call, for example a `getter`, and implement Java deserialization hooks to somehow inject the service instance.
+
+In Spring projects using Vaadin Kubernetes Kit, you can rely on transient field handling, and simply add the field for the service instance.
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:236]: And also here.
```suggestion
In Spring projects using Kubernetes Kit, you can rely on transient field handling, and simply add the field for the service instance.
```

File: articles/tools/azure/session-replication.adoc [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
+public class DataProviderUtil {
+  
+    // (1)
+	public static <S, T> T convertIfNotNull(S source, Function<S, T> converter, Supplier<T> nullValueSupplier) {
+ 		return source != null ? converter.apply(source) : nullValueSupplier.get();
+ 	}
+ 
+    // (2)
+	public static <T> ItemLabelGenerator<T> createItemLabelGenerator(Function<T, String> converter) {
+ 		return item -> convertIfNotNull(item, /* (3) */ converter, () -> "");
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:174]: Move `(3)` at the end of the line.
```suggestion
 		return item -> convertIfNotNull(item, converter, () -> ""); // (3)
```

File: articles/tools/azure/session-replication.adoc [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+order: 100
+---
+
+= Session Replication
+
+For a Vaadin application to reach the goals of High Availability and Down Scaling, it is fundamental that the HTTP session is correctly distributed between multiple servers.
+
+If the server processing user requests becomes unavailable, it should be transparently replaced by another one, in a way that the user should be able to continue using the application without any disruption.
+
+The user may notice a temporarily offline notification or the loading indicator spawn, but the browser page is not reloaded and work can continue without losing any data.
```
Comment (heruan) [inline articles/tools/azure/session-replication.adoc:13]: Rephrase this as discussed in the DX Test.

**Review by heruan — CHANGES_REQUESTED** [review]

File: articles/tools/azure/session-replication.adoc (hunk 14 above)
Comment (mcollovati) [reply articles/tools/azure/session-replication.adoc:13]: Proposed a change as review comment

File: articles/tools/azure/session-replication.adoc [hunk 15]
```diff
@@ -0,0 +1,551 @@
...
+Of course, after deserialization transient fields are null.
+
+For Spring applications, Kubernetes Kit is also able to overcome this limitation.
+Transient fields referencing Spring beans are detected during serialization and metadata is stored along with the session attributes.
+On deserialization, transient fields matching metadata are injected with the corresponding Spring bean.
+
+Transient field inspection and injection is performed using Java reflection.
+For performance reason and to avoid potential issues with reflection access checks, it is recommended to restrict the inspection to application classes only.
+
+Restrictions can be imposed by configuring `vaadin.serialization.include-packages` and `vaadin.serialization.exclude-packages` properties, to limit the inspection to classes whose packages matches the configured rules.
```
Comment (MatthewVaadin) [inline articles/tools/azure/session-replication.adoc:36]: Add a note explaining why `exclude-packages` would be used, maybe with an example.

File: articles/tools/azure/session-replication.adoc [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+The aim of the following sections is to give tips and best practices on how to code a fully serializable Vaadin applications.
+
+
+=== Use Serializable Object as UI Components Members
+
+When writing a component make sure that all its non-transient members implement [classname]`Serializable`.
+
+All Vaadin UI components and listeners are already marked as serializable, so if a class extends or implements one of them, it is not necessary to explicitly add the `Serializable` implementation statement.
+
+However, it is mandatory to do that on non UI components that may be referenced in the UI graph.
```
Comment (MatthewVaadin) [inline articles/tools/azure/session-replication.adoc:66]:
```suggestion
However, it is mandatory to do so on non UI components that may be referenced in the UI graph.
```

File: articles/tools/azure/session-replication.adoc [hunk 17]
```diff
@@ -0,0 +1,551 @@
...
+
+Once the lambda expression has been identified, replacing it with an anonymous class may be the solution.
+
+
+== Kubernetes Kit Session Replication Debug Tool
+
+To help making HTTP session fully serializable and deserializable, Kubernetes Kit offers a tool whose aim is to discover main issues during development.
+
+The debug tool is composed by a HTTP session replication tester, engaged by a Vaadin [classname]`RequestHandler`, that tries to serialize and deserialize the HTTP session for every user action on the browser that triggers a server request.
+It works only on `development mode` and only if session serialization is enabled setting the configuration property `vaadin.devmode.sessionSerialization.enabled` to `true`.
```
Comment (MatthewVaadin) [inline articles/tools/azure/session-replication.adoc:323]: Add a code block for the properties file.

File: articles/index.asciidoc [hunk 18]
```diff
@@ -135,6 +135,14 @@ Get insight into Flow applications at runtime and in production by monitoring ap
 [.sr-only]
 <<{articles}/tools/observability#,Observability Kit Overview>>
 
+[.card]
+=== Kubernetes Kit
+image::_images/kubernetes-kit.svg["", opts=inline, role=icon, width=48]
+Deploy your Vaadin applications to the cloud.
```
Comment (tarekoraby) [inline articles/index.asciidoc:141]:
```suggestion
Deploy your Vaadin application to Kubernetes, enabling scalability, high availability, and rolling updates.
```

File: articles/tools/index.asciidoc [hunk 19]
```diff
@@ -59,6 +59,14 @@ Get insight into Flow applications at runtime and in production by monitoring ap
 [.sr-only]
 <<{articles}/tools/observability#,See Observability Kit>>
 
+[.card]
+=== Kubernetes Kit
+image::{articles}/_images/kubernetes-kit.svg[opts=inline, role=icon]
+Deploy your Vaadin applications to the cloud.
```
Comment (tarekoraby) [inline articles/tools/index.asciidoc:65]:
```suggestion
Deploy your Vaadin application to Kubernetes, enabling scalability, high availability, and rolling updates.
```

File: articles/tools/kubernetes/index.asciidoc [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+= Kubernetes Kit
+
+:commercial-feature: Kubernetes Kit
+:kit-trial: true
+include::{articles}/_commercial-banner.asciidoc[opts=optional]
+
+Kubernetes is an open-source system for automating deployment, scaling, and management of containerized applications.
+
+The Kubernetes Kit is used to integrate your applications with Kubernetes, to be able to deploy your Vaadin Flow applications to the cloud.
```
Comment (tarekoraby) [inline articles/tools/kubernetes/index.asciidoc:15]:
```suggestion
The Kubernetes Kit is used to integrate your applications with Kubernetes to be able to deploy your Vaadin Flow applications on-premises or to the cloud.
```

File: articles/tools/kubernetes/index.asciidoc [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+The Kubernetes Kit enables teams to deploy seamlessly and operate Vaadin Java-based applications built using Flow. The Kit helps teams to quickly create Vaadin applications that are scalable, highly available, and user-friendly.
```
Comment (tarekoraby) [inline articles/tools/kubernetes/index.asciidoc:17]:
```suggestion
The Kubernetes Kit enables teams to deploy seamlessly and operate Vaadin Java-based applications built using Flow.
The Kit helps teams to quickly create Vaadin applications that are scalable, highly available, and user-friendly.
Specifically, the enables the following:
- Non-disruptive rolling updates that don't interrupt user sessions, reducing the cost and inconvenience of after-hours deployments. 
- Horizontal scalability, saving on cloud costs by allowing your application to scale down without impacting active users' sessions, and scale up when needed.
- High availability, enabling users to keep their active sessions and continue using your application even if a server fails.
- “serialization helpers” that make it faster and easier to fully leverage horizontal scaling and fail-over.
```

File: articles/tools/kubernetes/getting-started.asciidoc [hunk 22]
```diff
@@ -0,0 +1,338 @@
...
+description: Step-by-step guide showing how to add Kubernetes Kit
+  to your application.
+order: 10
+---
+
+= Getting Started with Kubernetes Kit
+:sectnums:
+
+Kubernetes Kit provides high availability for Vaadin applications
+ running in a Kubernetes cluster (including Azure).
```
Comment (tarekoraby) [inline articles/tools/kubernetes/getting-started.asciidoc:12]:
```suggestion
Kubernetes Kit enables scalability, high availability, and non-disruptive rolling updates for Vaadin applications running in a Kubernetes cluster.
```

**Review by tarekoraby — CHANGES_REQUESTED** [review]

**Review by tarekoraby — APPROVED** [review]

**Review by heruan — APPROVED** [review]
